src/catalog/src/durable/persist.rs

```diff
@@ -984,30 +1018,39 @@ impl UnopenedPersistCatalogState {
         self.mode = mode;
         let read_only = matches!(self.mode, Mode::Readonly);
 
-        self.sync_to_current_upper().await?;
-        let prev_epoch = self.epoch.validate()?;
         // Fence out previous catalogs.
-        let mut fence_updates = Vec::with_capacity(2);
-        if let Some(prev_epoch) = prev_epoch {
-            fence_updates.push((StateUpdateKind::Epoch(prev_epoch), -1));
-        }
-        let mut current_epoch = prev_epoch.unwrap_or(MIN_EPOCH).get();
-        // Only writable catalogs attempt to increment the epoch.
-        if matches!(self.mode, Mode::Writable) {
-            current_epoch = current_epoch + 1;
-        }
-        let current_epoch = Epoch::new(current_epoch).expect("known to be non-zero");
-        fence_updates.push((StateUpdateKind::Epoch(current_epoch), 1));
-        let current_epoch = FenceableToken::new(Some(current_epoch), FenceError::epoch);
-        debug!(
-            ?self.upper,
-            ?prev_epoch,
-            ?current_epoch,
-            "fencing previous catalogs"
-        );
-        self.epoch = current_epoch;
-        if matches!(self.mode, Mode::Writable) {
-            self.compare_and_append(fence_updates).await?;
+        loop {
```

Comment:
Should we have some upper bound on the number of times we retry? If we end up retrying more than say 10 times we should probably restart at some higher level (e.g. re-open the catalog or panic and restart the process)

Reply:
I'm not entirely convinced we should do this. One of the benefits of this change is to prevent callers from dealing with upper mismatches and having to retry themselves since we're almost always in a better position to do the retry more cheaply within the catalog. 

So the only thing we can do here on an upper mismatch is either retry ourself or panic. We could panic after 10 times or so, but if we ever encounter one of these panics, the only solution is to restart the process and try again which will eventually re-enter this loop after repeating a bunch of other work. Especially since we added the warnings, I almost think it would be better for an operator to manually kill the process if this is stuck and that's what they really want.

I might just be convincing myself that the path of less work is better so please push back if you disagree.

Reply:
+1 from me on this rationale—this feels a lot like when we taught persist to never return `Indeterminate` errors. In general infinite retries are scary, but I think they're appropriate in the components that can resolve indeterminate states. You've already got a log line in here, so if we do end up unexpectedly getting stuck in this infinite loop it will be immediately obvious where we're stuck.

If the log line is spammy because we regularly go through a turn of the loop on fencing, we could consider only emitting the log line on the second or third turn of the loop, but I think it's fine to just log unconditionally as you have here for the initial version of this.